File: BackEnd/SharedKernel/Aggregate/AggregateRoot.cs
namespace SharedKernel.Aggregate;

public abstract class AggregateRoot<TId> : IAggregateRoot where TId : notnull
{
    private readonly List<IDomainEvent> _domainEvents = [];
    
    public IReadOnlyCollection<IDomainEvent> DomainEvents => _domainEvents.AsReadOnly();
    
    public TId Id { get; set; }
    public DateTime? CreatedAt { get; set; }
    public Guid? CreatedBy { get; set; }
    public DateTime? UpdatedAt { get; set; }
    public Guid? UpdatedBy { get; set; }
    
    protected void AddDomainEvent(IDomainEvent domainEvent)
    {
        _domainEvents.Add(domainEvent);
    }

    public void ClearDomainEvents()
    {
        _domainEvents.Clear();
    }
    
}

File: flutter/lib/shared_kernel/aggregate/aggregate_root.dart
import 'package:enterprise_management/shared_kernel/domain/event/domain_event.dart';

abstract class AggregateRoot<TId> {
  AggregateRoot({required this.id, this.createdBy, this.createdAt, this.updatedBy, this.updatedAt});

  final TId id;
  final String? createdBy;
  final DateTime? createdAt;
  final String? updatedBy;
  final DateTime? updatedAt;

  final List<DomainEvent> _events = [];

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is AggregateRoot<TId> && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;

  List<DomainEvent> get events => List.unmodifiable(_events);

  void addEvent(DomainEvent event) {
    _events.add(event);
  }

  void addEvents(Iterable<DomainEvent> events) {
    _events.addAll(events);
  }

  void clearEvents() {
    _events.clear();
  }
}
